Sources/SentryCrash/Recording/Tools/SentryHook.c

```diff
@@ -45,17 +19,101 @@ sentrycrash__async_backtrace_incref(sentrycrash_async_backtrace_t *bt)
 }
 
 void
-sentrycrash__async_backtrace_decref(sentrycrash_async_backtrace_t *bt)
+sentrycrash_async_backtrace_decref(sentrycrash_async_backtrace_t *bt)
 {
     if (!bt) {
         return;
     }
     if (__atomic_fetch_add(&bt->refcount, -1, __ATOMIC_SEQ_CST) == 1) {
-        sentrycrash__async_backtrace_decref(bt->async_caller);
+        sentrycrash_async_backtrace_decref(bt->async_caller);
         free(bt);
     }
 }
 
+/**
+ * This is a poor-mans concurrent hashtable.
+ * We have N slots, using a simple FNV-like hashing function on the mach-thread id.
+ *
+ * *Writes* to a slot will only ever happen using the *current* thread. See
+ * the `set`/`unset` functions for SAFETY descriptions.
+ * *Reads* will mostly happen on the *current* thread as well, but might happen
+ * across threads through `sentrycrashsc_initWithMachineContext`. See the `get`
+ * function for possible UNSAFETY.
+ */
+
+#define SENTRY_MAX_ASYNC_THREADS (128 - 1)
+
+typedef struct {
+    SentryCrashThread thread;
+    sentrycrash_async_backtrace_t *backtrace;
+} sentrycrash_async_caller_slot_t;
+
+static sentrycrash_async_caller_slot_t sentry_async_callers[SENTRY_MAX_ASYNC_THREADS] = { 0 };
+
+static size_t
+sentrycrash__thread_idx(SentryCrashThread thread)
+{
+    return (thread * 0x100000001b3 + 0xcbf29ce484222325) % SENTRY_MAX_ASYNC_THREADS;
+}
+
+sentrycrash_async_backtrace_t *
+sentrycrash_get_async_caller_for_thread(SentryCrashThread thread)
+{
+    if (!__atomic_load_n(&hooks_active, __ATOMIC_RELAXED)) {
+        return NULL;
+    }
+
+    size_t idx = sentrycrash__thread_idx(thread);
+    sentrycrash_async_caller_slot_t *slot = &sentry_async_callers[idx];
+    if (__atomic_load_n(&slot->thread, __ATOMIC_ACQUIRE) == thread) {
+        sentrycrash_async_backtrace_t *backtrace
+            = __atomic_load_n(&slot->backtrace, __ATOMIC_RELAXED);
+        // UNSAFETY WARNING: There is a tiny chance of use-after-free here.
+        // This call can happen across threads, and the thread that "owns" the
+        // slot can decref and free the backtrace before *this* thread gets a
+        // chance to incref.
```

Comment:
maybe also mention that because of `suspendEnvironment` at call times we're risking this?